File: MMX-PromiseKitTests/MMX-PromiseKit.m
//
//  MMX-PromistKit.m
//  MMX-PromiseKit
//
//  Created by Seth on 8/26/15.
//  Copyright (c) 2015 Seth Samuel. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <PromiseKit/PromiseKit.h>
#import "MMX+PromiseKit.h"
#import <OCMock/OCMock.h>

@interface MMX_PromiseKit : XCTestCase

@end

@implementation MMX_PromiseKit

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

//MMXChannel

//+(PMKPromise*)channelsStartingWith:(NSString *)name limit:(int)limit;
- (void)testChannelsStartingWith {
    
    NSArray *stubChannels = @[@{},@{}];
    
    id mock = OCMClassMock([MMXChannel class]);
    [OCMStub([mock channelsStartingWith:[OCMArg any] limit:100 success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(int count, NSArray *channels) = nil;
        [invocation getArgument:&successBlock atIndex:4];
        successBlock(2, stubChannels);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"channelsStartingWith"];
    [MMXChannel channelsStartingWith:@"" limit:100]
    .then(^(NSNumber *count, NSArray* channels){
        XCTAssertEqual(count.integerValue, 2);
        XCTAssertNotNil(channels);
        XCTAssertEqualObjects(channels, stubChannels);

        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
      //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//+(PMKPromise*)findByTags:(NSSet*)tags;
- (void)testFindByTags {
    NSArray *stubChannels = @[@{},@{}];
    
    id mock = OCMClassMock([MMXChannel class]);
    [OCMStub([mock findByTags:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(int count, NSArray *channels) = nil;
        [invocation getArgument:&successBlock atIndex:3];
        successBlock(2, stubChannels);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"findByTags"];
    [MMXChannel findByTags:[NSSet setWithArray:@[@"foo"]]]
    .then(^(NSNumber *count, NSArray* channels){
        XCTAssertEqual(count.integerValue, 2);
        XCTAssertNotNil(channels);
        XCTAssertEqualObjects(channels, stubChannels);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)tags;
- (void)testTags {
    NSSet *stubTags = [NSSet setWithArray:@[@"foo"]];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock tagsWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(NSSet *tags) = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock(stubTags);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"tags"];
    [mock tags]
    .then(^(NSSet *tags){
        XCTAssertNotNil(tags);
        XCTAssertEqualObjects(tags, stubTags);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)setTags:(NSSet*)tags;
- (void)testSetTags {
    NSSet *stubTags = [NSSet setWithArray:@[@"foo"]];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock setTags:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:3];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"tags"];
    [mock setTags:stubTags]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)create;
- (void)testCreate {
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock createWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"create"];
    [mock create]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)delete;
- (void)testDelete {
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock deleteWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"delete"];
    [mock delete]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}


//-(PMKPromise*)subscribe;
- (void)testSubscribe {
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock subscribeWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"subscribe"];
    [mock subscribe]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}


//-(PMKPromise*)unSubscribe;
- (void)testUnSubscribe {
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock unSubscribeWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"unsubscribe"];
    [mock unSubscribe]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}


//+(PMKPromise*)subscribedChannels;
- (void)testSubscribedChannels {
    NSArray *stubChannels = @[@{},@{}];
    
    id mock = OCMClassMock([MMXChannel class]);
    [OCMStub([mock subscribedChannelsWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(NSArray *channels) = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock(stubChannels);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"subscribedChannels"];
    [MMXChannel subscribedChannels]
    .then(^(NSArray* channels){
        XCTAssertNotNil(channels);
        XCTAssertEqualObjects(channels, stubChannels);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];

}

//-(PMKPromise*)subscribers;
- (void)testSubscribers {
    NSArray *stubSubscribers = @[@{}];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock subscribersWithSuccess:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(int count, NSArray *subscribers) = nil;
        [invocation getArgument:&successBlock atIndex:2];
        successBlock(1, stubSubscribers);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"subscibers"];
    [mock subscribers]
    .then(^(NSNumber *count, NSArray *subscribers){
        XCTAssertEqual(count.integerValue, 1);
        XCTAssertNotNil(subscribers);
        XCTAssertEqualObjects(subscribers, stubSubscribers);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];

}

//-(PMKPromise*)publish:(NSDictionary *)messageContent;
- (void)testPublish {
    MMXMessage *stubMessage = [MMXMessage new];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock publish:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(MMXMessage *message) = nil;
        [invocation getArgument:&successBlock atIndex:3];
        successBlock(stubMessage);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"publish"];
    [mock publish:@{}]
    .then(^(MMXMessage *message){
        XCTAssertNotNil(message);
        XCTAssertEqualObjects(message, stubMessage);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)fetchMessagesBetweenStartDate:(NSDate *)startDate endDate:(NSDate *)endDate limit:(int)limit ascending:(BOOL)ascending;
- (void)testFetchMessagesBetweenStartDate {
    NSArray *stubMessages = @[@{},@{}];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock fetchMessagesBetweenStartDate:[OCMArg any] endDate:[OCMArg any] limit:100 ascending:YES success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(int totalCount, NSArray *messages) = nil;
        [invocation getArgument:&successBlock atIndex:6];
        successBlock(2, stubMessages);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"fetchMessagesBetweenStartDate"];
    [mock fetchMessagesBetweenStartDate:[NSDate date] endDate:[NSDate date] limit:100 ascending:YES]
    .then(^(NSNumber *totalCount, NSArray *messages){
        XCTAssertEqual(totalCount.integerValue, 2);
        XCTAssertNotNil(messages);
        XCTAssertEqualObjects(messages, stubMessages);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//-(PMKPromise*)inviteUser:(MMXUser *)user comments:(NSString *)comments;
- (void)testInviteUser {
    MMXInvite *stubInvite = [MMXInvite new];
    
    id mock = OCMPartialMock([MMXChannel new]);
    [OCMStub([mock inviteUser:[OCMArg any] comments:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)(MMXInvite *invite) = nil;
        [invocation getArgument:&successBlock atIndex:4];
        successBlock(stubInvite);
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"inviteUser"];
    [mock inviteUser:[MMXUser new] comments:@""]
    .then(^(MMXInvite *invite){
        XCTAssertNotNil(invite);
        XCTAssertEqualObjects(invite, stubInvite);
        
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
}

//MMXInvite
//-(PMKPromise*)acceptWithComments:(NSString *)comments;
- (void)testAcceptWithComments {

    id mock = OCMPartialMock([MMXInvite new]);
    [OCMStub([mock acceptWithComments:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:3];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"acceptWithComments"];
    [mock acceptWithComments:@"foo"]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];

}

//-(PMKPromise*)declineWithComments:(NSString *)comments;
- (void)testDeclineWithComments {
    
    id mock = OCMPartialMock([MMXInvite new]);
    [OCMStub([mock declineWithComments:[OCMArg any] success:[OCMArg any] failure:[OCMArg any]]) andDo:^(NSInvocation *invocation) {
        void (^successBlock)() = nil;
        [invocation getArgument:&successBlock atIndex:3];
        successBlock();
    }];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"declineWithComments"];
    [mock declineWithComments:@"foo"]
    .then(^(){
        [expectation fulfill];
    })
    .catch(^(NSError *error){
        XCTAssertNotNil(error);
        NSLog(@"%@", error);
        //Ignore any errors from authentication
    })
    .finally(^(){
    });
    
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        XCTAssertNil(error);
    }];
    
}


//
//MMXMessage
//-(PMKPromise*)send;
//-(PMKPromise*)replyWithContent:(NSDictionary *)content;
//-(PMKPromise*)replyAllWithContent:(NSDictionary *)content;
//
//MMXUser
//-(PMKPromise*)registerWithCredential:(NSURLCredential *)credential;
//+(PMKPromise*)logInWithCredential:(NSURLCredential *)credential;
//+(PMKPromise*)logOut;
//-(PMKPromise*)changePasswordWithCredential:(NSURLCredential *)credential;
//+(PMKPromise*)findByDisplayName:(NSString *)displayName limit:(int)limit;
//+(PMKPromise*)userForUsername:(NSString *)username;

@end

File: MMX-PromiseKitTests/MMX-PromiseKit.swift
//
//  MMX-PromiseKit.swift
//  MMX-PromiseKit
//
//  Created by Seth on 9/1/15.
//  Copyright (c) 2015 Seth Samuel. All rights reserved.
//

import Foundation
import MMX
import MMX_PromiseKit
import OCMock
import XCTest

private struct stubs{
    static let channels = ["channel"]
    static let tags = Set(["tag", ""])
    static let subscribers = ["user1"]
    static let messageContent : [String : AnyObject] = ["Type" : 1, "Message" : "Hello"]
    static let user = MMXUser()
}

class MMX_PromiseKit_MMXChannel : XCTestCase{

    func testFindByTags(){

        class MockMMXChannel : MMXChannel {
            override class func findByTags(tags: Set<NSObject>!, success: ((Int32, [AnyObject]!) -> Void)!, failure: ((NSError!) -> Void)!) -> Void{
                success(2, stubs.channels)
            }
            
        }

        let ex = expectationWithDescription("findByTags")

        MockMMXChannel.findByTags(["foo"])
            .then{ count, channels -> Void in
                XCTAssertEqual(count, 2, "Count equal")
                XCTAssertEqual(channels, stubs.channels, "Channels equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testChannelsStartingWithName() {
        class MockMMXChannel : MMXChannel {
            override class func channelsStartingWith(name: String!, limit : Int32, success: ((Int32, [AnyObject]!) -> Void)!, failure: ((NSError!) -> Void)!) -> Void{
                success(2, stubs.channels)
            }
            
        }
        
        let ex = expectationWithDescription("channelsStartingWithName")
        
        MockMMXChannel.channelsStartingWithName("foo", limit: 100)
            .then{ count, channels -> Void in
                XCTAssertEqual(count, 2, "Count equal")
                XCTAssertEqual(channels, stubs.channels, "Channels equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testTags(){
        class MockMMXChannel : MMXChannel {
            override func tagsWithSuccess(success: ((Set<NSObject>!) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(stubs.tags)
            }
            
        }
        
        let ex = expectationWithDescription("tags")
        
        MockMMXChannel().tags()
            .then{ tags -> Void in
                XCTAssertEqual(tags, stubs.tags, "Tags equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testSetTags(){
        class MockMMXChannel : MMXChannel {
            override func setTags(tags: Set<NSObject>!, success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success();
            }
        }
        
        let ex = expectationWithDescription("setTags")
        
        MockMMXChannel().setTags(Set(["newtag"]))
            .then{ 
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testCreate(){
        class MockMMXChannel : MMXChannel {
            override func createWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("create")
        
        MockMMXChannel().create()
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testDelete(){
        class MockMMXChannel : MMXChannel {
            override func deleteWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("delete")
        
        MockMMXChannel().delete()
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testSubscribe(){
        class MockMMXChannel : MMXChannel {
            override func subscribeWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("subscribe")
        
        MockMMXChannel().subscribe()
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testUnSubscribe(){
        class MockMMXChannel : MMXChannel {
            override func unSubscribeWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("unSubscribe")
        
        MockMMXChannel().unSubscribe()
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }
    
    func testSubscribedChannels(){
        class MockMMXChannel : MMXChannel {
            override class func subscribedChannelsWithSuccess(success: (([AnyObject]?) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(stubs.channels)
            }
        }
        
        let ex = expectationWithDescription("subscribedChannelsWithSuccess")
        
        MockMMXChannel.subscribedChannels()
            .then{ channels -> Void in
                XCTAssertEqual(channels, stubs.channels, "Channels equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testSubscribers(){
        class MockMMXChannel : MMXChannel {
            override func subscribersWithSuccess(success: ((Int32, [AnyObject]!) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(2, stubs.subscribers)
            }
        }
        
        let ex = expectationWithDescription("subscribers")
        
        MockMMXChannel().subscribers()
            .then{ count, subscribers -> Void in
                XCTAssertEqual(count, 2, "Count equal")
                XCTAssertEqual(subscribers, stubs.subscribers, "Subscribers equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testPublishMessageContent(){
        class MockMMXChannel : MMXChannel {
            override func publish(messageContent: [NSObject : AnyObject]!, success: ((MMXMessage!) -> Void)!, failure: ((NSError!) -> Void)!) {
                let message = MMXMessage()
                message.messageContent = messageContent
                success(message)
            }
        }
        
        let ex = expectationWithDescription("publish")

        
        MockMMXChannel().publishMessageContent(stubs.messageContent)
            .then{ message -> Void in
                let messageKeys = message.messageContent.keys.array
                let stubKeys = stubs.messageContent.keys.array
                XCTAssertEqual(messageKeys, stubKeys as [NSObject], "Keys equal");
                XCTAssertEqual(message.messageContent["Type"] as! Int, stubs.messageContent["Type"] as! Int, "Int value equal");                
                XCTAssertEqual(message.messageContent["Message"] as! String, stubs.messageContent["Message"] as! String, "String value equal");
                ex.fulfill()
        }
        waitForExpectationsWithTimeout(2, handler: nil)

    }
    
    func testFetchMessagesBetweenStartDate(){
        class MockMMXChannel : MMXChannel {
            override func fetchMessagesBetweenStartDate(startDate: NSDate!, endDate: NSDate!, limit: Int32, ascending: Bool, success: ((Int32, [AnyObject]!) -> Void)!, failure: ((NSError!) -> Void)!) {
                let message = MMXMessage()
                message.messageContent = stubs.messageContent
                success(1, [message])
            }
        }
        
        let ex = expectationWithDescription("fetchMessagesBetweenStartDate")
        
        
        MockMMXChannel().fetchMessagesBetweenStartDate(NSDate(), endDate: NSDate(), limit: 1, ascending: true)
            .then{ count, messages -> Void in
                XCTAssertEqual(count, 1, "Count equal")
                if let message = messages.first{
                    let messageKeys = message.messageContent.keys.array
                    let stubKeys = stubs.messageContent.keys.array
                    XCTAssertEqual(messageKeys, stubKeys as [NSObject], "Keys equal");
                    XCTAssertEqual(message.messageContent["Type"] as! Int, stubs.messageContent["Type"] as! Int, "Int value equal");
                    XCTAssertEqual(message.messageContent["Message"] as! String, stubs.messageContent["Message"] as! String, "String value equal");
                }else{
                    XCTAssert(false, "Message returned")
                }
                ex.fulfill()
        }
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }
    
    func testInviteUser(){
        class MockMMXChannel : MMXChannel {
            override func inviteUser(user: MMXUser!, comments: String!, success: ((MMXInvite!) -> Void)!, failure: ((NSError!) -> Void)!) -> String! {
                success(MMXInvite())
                return "1"
            }
        }
        
        let ex = expectationWithDescription("inviteUser")
        
        let (messageId, promise) = MockMMXChannel().inviteUser(MMXUser(), comments: "foo")
            promise
            .then{ invite -> Void in
                XCTAssertNotNil(invite, "Invite not nil")
                ex.fulfill()
        }
        XCTAssertEqual(messageId, "1", "Message id returned")
        waitForExpectationsWithTimeout(2, handler: nil)
    }
}

class MMX_PromiseKit_MMXInvite : XCTestCase{

    func testAcceptWithComments(){
        class MockMMXInvite : MMXInvite{
            override func acceptWithComments(comments: String!, success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        let ex = expectationWithDescription("acceptWithComments")
        
        MockMMXInvite().acceptWithComments("foo")
            .then{ invite -> Void in
                ex.fulfill()
        }
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }

    func testDeclineWithComments(){
        class MockMMXInvite : MMXInvite{
            override func declineWithComments(comments: String!, success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        let ex = expectationWithDescription("declineWithComments")
        
        MockMMXInvite().declineWithComments("foo")
            .then{ invite -> Void in
                ex.fulfill()
        }
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }
}

class MMX_PromiseKit_MMXMessage : XCTestCase{

    func testSend(){
        class MockMMXMessage : MMXMessage{
            override func sendWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) -> String! {
                success()
                return "1"
            }
        }
        
        let ex = expectationWithDescription("send")
        
        let (messageId, promise) = MockMMXMessage().send()
        promise
            .then{
                ex.fulfill()
        }
        XCTAssertEqual(messageId, "1", "Message id returned")
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testReplyWithContent(){
        class MockMMXMessage : MMXMessage{
            override func replyWithContent(content : [NSObject:AnyObject], success: (() -> Void)!, failure: ((NSError!) -> Void)!) -> String! {
                success()
                return "1"
            }
        }
        
        let ex = expectationWithDescription("replyWithContent")
        
        let (messageId, promise) = MockMMXMessage().replyWithContent(["foo":"bar"])
        promise
            .then{
                ex.fulfill()
        }
        XCTAssertEqual(messageId, "1", "Message id returned")
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testReplyAllWithContent(){
        class MockMMXMessage : MMXMessage{
            override func replyAllWithContent(content : [NSObject:AnyObject], success: (() -> Void)!, failure: ((NSError!) -> Void)!) -> String! {
                success()
                return "1"
            }
        }
        
        let ex = expectationWithDescription("replyAllWithContent")
        
        let (messageId, promise) = MockMMXMessage().replyAllWithContent(["foo":"bar"])
        promise
            .then{
                ex.fulfill()
        }
        XCTAssertEqual(messageId, "1", "Message id returned")
        waitForExpectationsWithTimeout(2, handler: nil)
    }
}

class MMX_PromiseKit_MMXUser : XCTestCase{
    func testRegisterWithCredential(){
        class MockMMXUser : MMXUser{
            override func registerWithCredential(credential: NSURLCredential!, success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("registerWithCredential")
        
        MockMMXUser().registerWithCredential(NSURLCredential())
            .then{
                ex.fulfill()
        }

        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testLogInWithCredential(){
        class MockMMXUser : MMXUser{
            override class func logInWithCredential(credential: NSURLCredential!, success: ((MMXUser) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(stubs.user)
            }
        }
        
        let ex = expectationWithDescription("logInWithCredential")
        
        MockMMXUser.logInWithCredential(NSURLCredential())
            .then{ user -> Void in
                XCTAssertEqual(user, stubs.user, "User returned")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testLogout(){
        class MockMMXUser : MMXUser{
            override class func logOutWithSuccess(success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("logOut")
        
        MockMMXUser.logOut()
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testChangePasswordWithCredential(){
        class MockMMXUser : MMXUser{
            override func changePasswordWithCredential(credential: NSURLCredential!, success: (() -> Void)!, failure: ((NSError!) -> Void)!) {
                success()
            }
        }
        
        let ex = expectationWithDescription("changePasswordWithCredential")
        
        MockMMXUser().changePasswordWithCredential(NSURLCredential())
            .then{
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
    }
    
    func testFindByDisplayName(){
        class MockMMXUser : MMXUser {
            override class func findByDisplayName(displayName : String, limit : Int32, success: ((Int32, [AnyObject]!) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(2, [stubs.user])
            }
        }
        
        let ex = expectationWithDescription("findByDisplayName")
        
        MockMMXUser.findByDisplayName("name", limit: 10)
            .then{ count, users -> Void in
                XCTAssertEqual(count, 2, "Count equal")
                XCTAssertEqual(users, [stubs.user], "Users equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }
    
    func testUserForUsername(){
        class MockMMXUser : MMXUser {
            override class func userForUsername(username : String, success: ((MMXUser!) -> Void)!, failure: ((NSError!) -> Void)!) {
                success(stubs.user)
            }
        }
        
        let ex = expectationWithDescription("userForUsername")
        
        MockMMXUser.userForUsername("name")
            .then{ user -> Void in
                XCTAssertEqual(user, stubs.user, "User equal")
                ex.fulfill()
        }
        
        waitForExpectationsWithTimeout(2, handler: nil)
        
    }
}